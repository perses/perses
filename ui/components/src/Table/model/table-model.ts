// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Theme } from '@mui/material';
import {
  AccessorKeyColumnDef,
  CellContext,
  ColumnDef,
  CoreOptions,
  RowData,
  RowSelectionState,
  SortingState,
} from '@tanstack/react-table';
import { CSSProperties } from 'react';

export type TableDensity = 'compact' | 'standard';
export type SortDirection = 'asc' | 'desc' | undefined;

export type TableRowEventOpts = {
  /**
   * Unique identifier for the row.
   */
  id: string;

  /**
   * Index of the row in the original data.
   */
  index: number;
};

export interface TableProps<TableData> {
  /**
   * Height of the table.
   */
  height: number;

  /**
   * Width of the table.
   */
  width: number;

  /**
   * Array of data to render in the table. Each entry in the array will be
   * rendered as a row in the table.
   */
  data: TableData[];

  /**
   * Array of column configuration for the table. Each entry in the array will
   * be rendered as a column header and impact the rendering of cells within
   * table rows.
   */
  columns: Array<TableColumnConfig<TableData>>;

  /**
   * The density of the table layout. This impacts the size and space taken up
   * by content in the table (e.g. font size, padding).
   */
  density?: TableDensity;

  /**
   * When `true`, the first column of the table will include checkboxes.
   */
  checkboxSelection?: boolean;

  /**
   * Determines the behavior of row selection.
   *
   * - `standard`: clicking a checkbox will toggle that rows's selected/unselected
   *    state and will not impact other rows.
   * - `legend`: clicking a checkbox will "focus" that row by selecting it and
   *    unselecting other rows. Clicking a checkbox with a modifier key pressed,
   *    will change this behavior to behave like `standard`.
   *
   * @default 'standard'
   */
  rowSelectionVariant?: 'standard' | 'legend';

  /**
   * State of selected rows in the table when `checkboxSelection` is enabled.
   *
   * Selected row state is a controlled value that should be managed using a
   * combination of this prop and `onRowSelectionChange`.
   */
  rowSelection?: RowSelectionState;

  /**
   * Callback fired when the mouse is moved over a table row.
   */
  onRowMouseOver?: (e: React.MouseEvent, opts: TableRowEventOpts) => void;

  /**
   * Callback fired when the mouse is moved out of a table row.
   */
  onRowMouseOut?: (e: React.MouseEvent, opts: TableRowEventOpts) => void;

  /**
   * State of the column sorting in the table.
   *
   * The column sorting is a controlled value that should be managed using a
   * combination fo this prop and `onSortingChange`.
   */
  sorting?: SortingState;

  /**
   * Callback fired when the selected rows in the table changes.
   */
  onRowSelectionChange?: (rowSelection: RowSelectionState) => void;

  /**
   * Callback fired when the table sorting changes.
   */
  onSortingChange?: (sorting: SortingState) => void;

  /**
   * Function used to determine the unique identifier used for each row. This
   * value is used to key `rowSelection`. If this value is not set, the index
   * is used as the unique identifier.
   */
  getRowId?: CoreOptions<TableData>['getRowId'];

  /**
   * Function used to determine the color of the checkbox when `checkboxSelection`
   * is enabled. If not set, a default color is used.
   */
  getCheckboxColor?: (rowData: TableData) => string;
}

function calculateTableCellHeight(lineHeight: CSSProperties['lineHeight'], paddingY: string): number {
  // Doing a bunch of math to enforce height to avoid weirdness with mismatched
  // heights based on customization of cell contents.
  const lineHeightNum = typeof lineHeight === 'string' ? parseInt(lineHeight, 10) : lineHeight ?? 0;
  const verticalPaddingNum = typeof paddingY === 'string' ? parseInt(paddingY, 10) : paddingY;

  return lineHeightNum + verticalPaddingNum * 2;
}

type TableCellLayout = NonNullable<Pick<React.CSSProperties, 'padding' | 'fontSize' | 'lineHeight'>> & {
  height: number;
};

type GetTableCellLayoutOpts = {
  isLastColumn?: boolean;
  isFirstColumn?: boolean;
};

/**
 * Returns the properties to lay out the content of table cells based on the
 * theme and density.
 */
export function getTableCellLayout(
  theme: Theme,
  density: TableDensity,
  { isLastColumn, isFirstColumn }: GetTableCellLayoutOpts = {}
): TableCellLayout {
  if (density === 'compact') {
    const paddingY = theme.spacing(0.5);

    const basePaddingX = theme.spacing(0.5);
    const edgePaddingX = theme.spacing(1);
    const paddingLeft = isFirstColumn ? edgePaddingX : basePaddingX;
    const paddingRight = isLastColumn ? edgePaddingX : basePaddingX;

    const lineHeight = theme.typography.body2.lineHeight;

    return {
      padding: `${paddingY} ${paddingRight} ${paddingY} ${paddingLeft}`,
      height: calculateTableCellHeight(lineHeight, paddingY),
      fontSize: theme.typography.body2.fontSize,
      lineHeight: lineHeight,
    };
  }

  // standard density
  const paddingY = theme.spacing(1);
  const basePaddingX = theme.spacing(1.25);
  const edgePaddingX = theme.spacing(2);
  const paddingLeft = isFirstColumn ? edgePaddingX : basePaddingX;
  const paddingRight = isLastColumn ? edgePaddingX : basePaddingX;
  const lineHeight = theme.typography.body1.lineHeight;

  return {
    padding: `${paddingY} ${paddingRight} ${paddingY} ${paddingLeft}`,
    height: calculateTableCellHeight(lineHeight, paddingY),
    fontSize: theme.typography.body1.fontSize,
    lineHeight: lineHeight,
  };
}

export type TableCellAlignment = 'left' | 'right' | 'center';

// Overrides of meta value, so it can have a meaningful type in our code instead
// of `any`. Putting this in the model instead of a separate .d.ts file because
// I couldn't get it to work properly that way and punted on figuring it out
// after trying several things.
declare module '@tanstack/table-core' {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  interface ColumnMeta<TData extends RowData, TValue> {
    align?: TableColumnConfig<TData>['align'];
    headerDescription?: TableColumnConfig<TData>['headerDescription'];
    cellDescription?: TableColumnConfig<TData>['cellDescription'];
  }
}

// Only exposing a very simplified version of the very extensive column definitions
// possible with tanstack table to make it easier for us to control rendering
// and functionality.
export interface TableColumnConfig<TableData>
  // Any needed to work around some typing issues with tanstack query.
  // https://github.com/TanStack/table/issues/4241
  // TODO: revisit issue thread and see if there are any workarounds we can
  // use.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  extends Pick<AccessorKeyColumnDef<TableData, any>, 'accessorKey' | 'cell' | 'sortingFn'> {
  /**
   * Text to display in the header for the column.
   */
  header: string;

  /**
   * Alignment of the content in the cell.
   */
  align?: TableCellAlignment;

  /**
   * Text to display when hovering over a cell. This can be useful for
   * providing additional information about the column when the content is
   * ellipsized to fit in the space.
   *
   * If set to `true`, it will use the value generated by the `cell` prop if it
   * can be treated as a string.
   */
  // `any` needed for same reason as no-explicit-any higher up in this type.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  cellDescription?: ((props: CellContext<TableData, any>) => string) | boolean | undefined;

  /**
   * When `true`, the column will be sortable.
   * @default false
   */
  enableSorting?: boolean;

  /**
   * Text to display when hovering over the header text. This can be useful for
   * providing additional information about the column when you want to keep the
   * header text relatively short to manage the column width.
   */
  headerDescription?: string;

  // Tanstack Table does not support an "auto" value to naturally size to fit
  // the space in a table. Adding a custom setting to manage this ourselves.
  /**
   * Width of the column when rendered in a table. It should be a number in pixels
   * or "auto" to allow the table to automatically adjust the width to fill
   * space.
   * @default 'auto'
   */
  width?: number | 'auto';
}

/**
 * Takes in a perses table column and transforms it into a tanstack column.
 */
export function persesColumnsToTanstackColumns<TableData>(columns: Array<TableColumnConfig<TableData>>) {
  const tableCols: Array<ColumnDef<TableData>> = columns.map(
    ({ width, align, headerDescription, cellDescription, enableSorting, ...otherProps }) => {
      // Tanstack Table does not support an "auto" value to naturally size to fit
      // the space in a table. We translate our custom "auto" setting to 0 size
      // for these columns, so it is easy to fall back to auto when rendering.
      // Taking from a recommendation in this github discussion:
      // https://github.com/TanStack/table/discussions/4179#discussioncomment-3631326
      const sizeProps =
        width === 'auto' || width === undefined
          ? {
              // All zero values are used as shorthand for "auto" when rendering
              // because it makes it easy to fall back. (e.g. `row.size || "auto"`)
              size: 0,
              minSize: 0,
              maxSize: 0,
            }
          : {
              size: width,
            };

      const result = {
        ...otherProps,
        ...sizeProps,

        // Default sorting to false, so it is very explicitly set per column.
        enableSorting: !!enableSorting,

        // Open-ended store for extra metadata in TanStack Table, so you can bake
        // in your own features.
        meta: {
          align,
          headerDescription,
          cellDescription,
        },
      };

      return result;
    }
  );

  return tableCols;
}
