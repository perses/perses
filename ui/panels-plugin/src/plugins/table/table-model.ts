// Copyright 2024 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Definition, Transform } from '@perses-dev/core';
import { TableDensity } from '@perses-dev/components';

export interface ColumnSettings {
  name: string;

  // Text to display in the header for the column.
  header?: string;
  /**
   * Text to display when hovering over the header text. This can be useful for
   * providing additional information about the column when you want to keep the
   * header text relatively short to manage the column width.
   */
  headerDescription?: string;
  /**
   * Text to display when hovering over a cell. This can be useful for
   * providing additional information about the column when the content is
   * ellipsized to fit in the space.
   */
  cellDescription?: string;
  // Alignment of the content in the cell.
  align?: 'left' | 'center' | 'right';

  // When `true`, the column will be sortable.
  enableSorting?: boolean;

  // TODO: Allow user to set default sort => sort?: 'asc' | 'desc';

  /**
   * Width of the column when rendered in a table. It should be a number in pixels
   * or "auto" to allow the table to automatically adjust the width to fill
   * space.
   */
  width?: number | 'auto';
  // When `true`, the column will not be displayed.
  hide?: boolean;
}

export interface ValueCondition {
  kind: 'Value';
  spec: {
    value: string;
  };
}

export interface RangeCondition {
  kind: 'Range';
  spec: {
    min?: number;
    max?: number;
  };
}

export interface RegexCondition {
  kind: 'Regex';
  spec: {
    expr: string;
  };
}

export interface MiscCondition {
  kind: 'Misc';
  spec: {
    value: 'empty' | 'null' | 'NaN' | 'true' | 'false';
  };
}

export type Condition = ValueCondition | RangeCondition | RegexCondition | MiscCondition;

export interface CellSettings {
  condition: Condition;
  text?: string;
  textColor?: `#${string}`;
  backgroundColor?: `#${string}`;
}

/**
 * The schema for a Table panel.
 */
export interface TableDefinition extends Definition<TableOptions> {
  kind: 'Table';
}

/**
 * The Options object type supported by the Table panel plugin.
 */
export interface TableOptions {
  density?: TableDensity;
  columnSettings?: ColumnSettings[];
  cellSettings?: CellSettings[];
  transforms?: Transform[];
}

/**
 * Creates the initial/empty options for a Table panel.
 */
export function createInitialTableOptions(): TableOptions {
  return {
    density: 'standard',
  };
}
