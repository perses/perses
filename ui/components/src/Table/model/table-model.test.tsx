// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createTheme } from '@mui/material';
import { TableColumnConfig, getTableCellLayout, persesColumnsToTanstackColumns } from './table-model';

const mockMuiTheme = createTheme({});

describe('getTableCellLayout', () => {
  test.each(['compact', 'standard'] as const)('gets layout for %s density', (density) => {
    expect(getTableCellLayout(mockMuiTheme, density)).toMatchSnapshot();
  });
});

type MockTableData = {
  id: string;
  label: string;
  value: number;
};

describe('persesColumnToTanstackColumn', () => {
  test('maps "auto" width to all zero `size` values', () => {
    const persesColumns: Array<TableColumnConfig<MockTableData>> = [
      {
        accessorKey: 'label',
        header: 'Name',
        width: 'auto',
      },
    ];
    const tanstackColumns = persesColumnsToTanstackColumns(persesColumns);
    expect(tanstackColumns[0]).toEqual(
      expect.objectContaining({
        size: 0,
        minSize: 0,
        maxSize: 0,
      })
    );
  });

  test('persists numeric width value as `size`', () => {
    const persesColumns: Array<TableColumnConfig<MockTableData>> = [
      {
        accessorKey: 'label',
        header: 'Name',
        width: 100,
      },
    ];
    const tanstackColumns = persesColumnsToTanstackColumns(persesColumns);
    expect(tanstackColumns[0]).toEqual(
      expect.objectContaining({
        size: 100,
      })
    );
  });

  test('transforms perses columns to tanstack columns', () => {
    const persesColumns: Array<TableColumnConfig<MockTableData>> = [
      {
        accessorKey: 'label',
        header: 'Name',
        width: 'auto',
      },
      {
        accessorKey: 'value',
        header: 'Count',
        width: 120,
        cell: (data) => <strong>{data.getValue()}</strong>,
      },
    ];
    expect(persesColumnsToTanstackColumns(persesColumns)).toMatchSnapshot();
  });
});
