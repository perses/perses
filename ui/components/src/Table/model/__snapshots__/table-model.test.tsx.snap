// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTableCellLayout gets layout for compact density in center column 1`] = `
Object {
  "fontSize": "0.875rem",
  "height": 9.43,
  "lineHeight": 1.43,
  "padding": "4px 4px 4px 4px",
}
`;

exports[`getTableCellLayout gets layout for compact density in first column 1`] = `
Object {
  "fontSize": "0.875rem",
  "height": 9.43,
  "lineHeight": 1.43,
  "padding": "4px 4px 4px 8px",
}
`;

exports[`getTableCellLayout gets layout for compact density in last column 1`] = `
Object {
  "fontSize": "0.875rem",
  "height": 9.43,
  "lineHeight": 1.43,
  "padding": "4px 8px 4px 4px",
}
`;

exports[`getTableCellLayout gets layout for standard density in center column 1`] = `
Object {
  "fontSize": "1rem",
  "height": 17.5,
  "lineHeight": 1.5,
  "padding": "8px 10px 8px 10px",
}
`;

exports[`getTableCellLayout gets layout for standard density in first column 1`] = `
Object {
  "fontSize": "1rem",
  "height": 17.5,
  "lineHeight": 1.5,
  "padding": "8px 10px 8px 16px",
}
`;

exports[`getTableCellLayout gets layout for standard density in last column 1`] = `
Object {
  "fontSize": "1rem",
  "height": 17.5,
  "lineHeight": 1.5,
  "padding": "8px 16px 8px 10px",
}
`;

exports[`persesColumnToTanstackColumn transforms perses columns to tanstack columns 1`] = `
Array [
  Object {
    "accessorKey": "label",
    "header": "Name",
    "maxSize": 0,
    "meta": Object {
      "align": "right",
      "cellDescription": undefined,
      "headerDescription": undefined,
    },
    "minSize": 0,
    "size": 0,
  },
  Object {
    "accessorKey": "value",
    "cell": [Function],
    "header": "Count",
    "meta": Object {
      "align": undefined,
      "cellDescription": [Function],
      "headerDescription": "The total number of values.",
    },
    "size": 120,
  },
]
`;
